src/ConcatenateMsg.chpl

```diff
@@ -653,4 +653,100 @@ module ConcatenateMsg
         return new MsgTuple(repMsg, MsgType.NORMAL);
     }
     registerFunction("concatenateUniquely", concatenateUniqueStrMsg, getModuleName());
+
+    proc concatenateUniqueStrMsg2(cmd: string, msgArgs: borrowed MessageArgs, st: borrowed SymTab): MsgTuple throws {
+        param pn = Reflection.getRoutineName();
+
+        var repMsg: string;
+        var names = msgArgs.get("names").toScalarList(string);
+        var n = names.size;
+
+
+        cmLogger.debug(getModuleName(), getRoutineName(), getLineNumber(),
+                    "concatenate unique strings from %i arrays: %?".format(n, names));
+
+        // Each locale gets its own set
+        var localeSets = makeDistArray(numLocales, set(string));
+
+        // Initialize sets
+        coforall loc in Locales do on loc {
+            localeSets[here.id] = new set(string);
+        }
+
+        var stringsInLocale: [PrivateSpace] innerArray(string);
+
+        // Collect all unique strings from each input SegmentedString
+        for i in 0..#names.size {
+            var rawName = names[i];
+            var (strName, _) = rawName.splitMsgToTuple('+', 2);
+            try {
+                var segString = getSegString(strName, st);
+                cmLogger.debug(getModuleName(), getRoutineName(), getLineNumber(),
+                            "Processing SegString: %s".format(strName));
+
+                // Grab the strings by locale and throw them in the sets.
+                var stringArrays = balancedSegStringRetrieval(segString);
+                coforall loc in Locales do on loc {
+                    ref myStrings = stringArrays[here.id].Arr;
+                    var locSet = new set(string);
+                    
+                    // Reduce to unique strings within this locale
+                    forall str in myStrings with (+ reduce locSet) {
+                        locSet.add(str);
+                    }
```

Comment:
This is an interesting pattern!

It looks like in some tests, at least, using `set(string, parSafe=true)` instead and not doing things through the `+ reduce` intent results in faster performance.